sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala

```diff
@@ -451,10 +451,25 @@ class SymmetricHashJoinStateManager(
   }
 
   private trait KeyWithIndexToValueRowConverter {
+    /** Defines the schema of the value row (the value side of K-V in state store). */
     def valueAttributes: Seq[Attribute]
 
+    /**
+     * Convert the value row to (actual value, match) pair.
+     *
+     * NOTE: implementations should ensure the result row is NOT reused during execution, as
+     * caller may use the value to store without copy().
```

Comment:
I don't know the details, but how likely the caller side will store the values?

Reply:
So the problem occurred here:

https://github.com/apache/spark/blob/42f01e314b4874236544cc8b94bef766269385ee/sql/core/src/main/scala/org/apache/spark/sql/execution/streaming/state/SymmetricHashJoinStateManager.scala#L252-L281

Here we temporarily store `currentValue` via calling `convertValue` which is candidate to remove. After that we adjust the array a bit to not having a hole in between, where it also triggers `convertValue` and "changes" the value in `currentValue` silently if we don't copy.

It's also possible for caller to call copy() but the row to copy is inside of ValueAndMatchPair so I feel it's pretty much easier to miss even with method doc, so I took safer approach rather than performance-wise.

Reply:
I don't have a strong opinion here. Just for reference: The SMJ copies the input row on-demand: only copy the rows when we need to buffer them. See `SortMergeJoinExec.bufferMatchingRows`. It's only a few places we need to buffer the rows, so it's OK to do it in SMJ. I'm not sure about streaming join though.

Reply:
For `KeyWithIndexToValueRowConverterFormatV1`, we don't have this issue because the `covertToValueRow` didn't change the `currentValue`?

Reply:
Yes. KeyWithIndexToValueRowConverterFormatV1 simply uses the row state store will provide, which is expected to be unique instance (as we copy it when putting to state store).

I've had time to go through callers, and I'm still hesitated to let callers deal with copy, because the row to take care is inside of the ValueAndMatchPair, and some of current callers are even re-packaging it to other class, e.g. KeyWithIndexAndValue for getAll(), so not only direct callers should take care of it but caller of the caller may have to take care of it. It would bring too much complication.

Reply:
I recall in Spark SQL, there are some places similar and usually we make it as the caller's responsibility to call copy. Is this a code path critical for performance?

Reply:
Looks like we'd prefer to open maximizing performance for all callers. I changed to not copy whenever possible, with adding NOTE in method comments on all impacting methods. be34258

Reply:
The disadvantage for the approach is, given the contract is applied on the interface level, row from format V1 now would get unnecessary copy whereas it doesn't need to. It only applies to the evicted rows and ideally end users would need to go with format V2, hope it is OK.